Do not follow symlinks in getting/setting file attributes

rlaboiss opened the pull request:
This is achieved in a quite convoluted way, but which works in Python2.
Note that an equivalent in Python3 would be much simpler, because all
the involved os.\* functions accept the argument 'follow_symlink=False'.

Here is the description of the changes in this commit:

1) Call os.lstat() for getting the file attributes.

2) Use os.lchown() for changing the gid and uid of the file.

3) If the file is a symbolic link, then does not call os.chmod() on it,
since it is not possible to change permissions of symbolic links (at
least in Linux).

4) Since there is no "os.lutime()" in the Python os module, the setting
of mtime is done done by calling of the touch system utility, with the
appropriate option "-t".  This function is widely available, so there
should be no problem in introducing this call.  However, in order to not
follow symlinks when touching, option "-h" must be used.  This option
appeared in version 8.1 of GNU coreutils.  In order to avoid errors, the
version of touch in the system is tested and he symbolic link's mtime is
adjusted only if GNU coureutils' touch is used AND if its version is >=
8.1. (Note that, for this to work, the distutils.version module is
imported).  Furthermore, the format of the date accepted by option -t of
the touch command is "YYYYmmddHHMM.SS".  Functions time.strftime() and
time.gmtime() are used for converting from the format returned by
os.lstat().  Note that the mtime is converted to GMT.

rlaboiss committed:
Do not follow symlinks in getting/setting file attributes

This is achieved in a quite convoluted way, but which works in Python2.
Note that an equivalent in Python3 would be much simpler, because all
the involved os.* functions accept the argument 'follow_symlink=False'.

Here is the description of the changes in this commit:

1) Call os.lstat() for getting the file attributes.

2) Use os.lchown() for changing the gid and uid of the file.

3) If the file is a symbolic link, then does not call os.chmod() on it,
since it is not possible to change permissions of symbolic links (at
least in Linux).

4) Since there is no "os.lutime()" in the Python os module, the setting
of mtime is done done by calling of the touch system utility, with the
appropriate option "-t".  This function is widely available, so there
should be no problem in introducing this call.  However, in order to not
follow symlinks when touching, option "-h" must be used.  This option
appeared in version 8.1 of GNU coreutils.  In order to avoid errors, the
version of touch in the system is tested and he symbolic link's mtime is
adjusted only if GNU coureutils' touch is used AND if its version is >=
8.1. (Note that, for this to work, the distutils.version module is
imported).  Furthermore, the format of the date accepted by option -t of
the touch command is "YYYYmmddHHMM.SS".  Functions time.strftime() and
time.gmtime() are used for converting from the format returned by
os.lstat().  Note that the mtime is converted to GMT.

renard commented:
Ok I see your point. Sorry for previous misguidance.

I guess this won't work on OSX (and other BSD system) since touch does not support `--version` BUT supports `-h`. I don't know about MS either.

Maybe a use of `sys.platform` should be a part of the test. And maybe a `--follow-symlink` option could take place to work around that feature.

rlaboiss committed:
Only check fot -h option of the touch command if the OS is Linux

In OSX (darwin), always use the -h option.  If it is not Linux neither
OSX, then the use of the touch command is disabled.

rlaboiss commented:
Ok, I added a second commit to this pull request, where sys.platform is used for testing if we are in Linux, in which case the version of touch is probed via --version. If we are in OSX, then option -h of touch is supposed to work. Otherwise, the use of touch is disabled.The version of git-meta in this branch should work well across the different platforms, I hope.

renard commented:
Merged. Thanks.
